### PR TITLE
Enable request body processing in detection mode in Nginx

### PR DESCRIPTION
**Pull request by vkrivopalov**
## Description

The driver for this change is a bug in the current behavior in detection-only mode that causes request body to be skipped on processing.

In order to address it, a more fundamental change has been applied to processing tasks spawned for asynchronous execution in detection-only mode. Instead of copying all the data for processing using Apache-specific memory pools, we now rely on internal Nginx reference counting for HTTP requests making them live long enough for processing to complete.

This allows us to save on memory allocations as there is no longer a need to copy request headers. But more importantly, this guarantees that stored request body buffers live long enough to be processed without opening a window for data races. This approach is considered the optimal one because request body can be stored in temporary files and maintaining their lifetime or copying files around is both more heavyweight and error-prone.

## Testing

All the listed tests are performed on an AddressSanitizer-instrumented debug binary of Nginx that terminates the process immediately on catching any memory issue. Logs have been tracked to seek for reported issues as well as processes to not be terminated.

All tests listed below are performed on each of the two following configurations:
 - WAF in detection-only mode, request body processing enabled
 - WAF in prevention mode, request body processing enabled

1.  `curl -X POST --header "Content-Type: application/json" --data "{\"value\": \"/etc/passwd\"}" "http://localhost/" -o /dev/null -s -w "got HTTP %{http_code} after %{time_total}s\n"`
    WAF messages for `/etc/passwd` appear in `waf_json.log`, no ASan-reported issues.
2. `curl -X POST --header "Content-Type: application/json" --data "{\"value\": \"/etc/passwd\"}" "http://localhost/" -o /dev/null -s -w "got HTTP %{http_code} after %{time_total}s\n"`
    WAF messages for weird sequences of characters appear in `waf_json.log`, no ASan-reported issues.
3. (From two terminals simultaneously) `./apib --input-file ./jsonpayload --method POST -H "Connection: Keep-Alive" -H "Accept: */*" --concurrency 10 --duration 600 http://localhost/`
 where `jsonpayload` contains `{"name": "John", "age": 31, "city": "New York"}`
  No ASan-reported errors.
4. `for i in `seq 1 50`; curl -X POST --data "$(openssl rand -hex 65535)" "http://localhost/" -o /dev/null -s -w "got HTTP %{http_code} after %{time_total}s\n" done`
  No ASan-reported errors.
5. Manual testing with `testbackend1.westus.cloudapp.azure.com` and `testbackend2.westus.cloudapp.azure.com` backends and various typical WAF options like `a=1=1` and `test=/etc/passwd` in query parameters and body. No issues discovered.

### CODE REVIEW

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 1]
```diff
@@ -761,53 +715,41 @@ static void
...
-    conn_rec *connection = ctx->connection;
-
-    if (req != NULL) {
-        modsecFinishRequest(req);
-    }
-    if (connection != NULL) {
-        modsecFinishConnection(connection);
-    }
+    ngx_http_request_t *r = ev->data;
+    --r->main->blocked;
```
Comment (azhao155) [inline nginx/modsecurity/ngx_http_modsecurity.c:741]: maybe add comments where it got incremented, so easy to pair them

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 2]
```diff
@@ -819,57 +761,72 @@ ngx_http_modsecurity_prevention_task_offload(ngx_http_request_t *r)
...
+    /* Increment request reference count to make sure it lives long enough
+     * for the task to be completed.
+     * Note that it is safe to do that after posting the task to the thread pool because
+     * we decrement it in the completion handler and it will only run after the current code
+     * as it is scheduled on the same (main) thread.
+     * Besides the reference counter ('count') we need to increment 'blocked' because this
+     * guarantees that the request is not terminated when the connection is closed -
+     * see ngx_http_terminate_handler() for how it neglects the 'count' value in this case.
+     */
+    ++r->main->count;
```
Comment (azhao155) [inline nginx/modsecurity/ngx_http_modsecurity.c:829]: in the completion, we only decrease block, so where the count got decremented. 
and i think count and block are used to control the lifecycle  of different struct, count is for the ctx and the block is for the request, correct me if i am wrong, so right now both ctx and request will be released at the same time.

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 3]
```diff
@@ -311,18 +273,18 @@ ngx_http_modsecurity_load_request(ngx_http_request_t *r)
         return NGX_ERROR;
     }
 
-    req->filename = dup_ngx_str_to_apr(req->pool, &path);
+    req->filename = (char *) path.data;
```
Comment (allanrbo) [inline nginx/modsecurity/ngx_http_modsecurity.c:276]: are we sure the data in this ngx_str_t is null terminated?

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 4]
```diff
@@ -761,53 +715,41 @@ static void
...
-
-    if (req != NULL) {
-        modsecFinishRequest(req);
-    }
-    if (connection != NULL) {
-        modsecFinishConnection(connection);
-    }
+    ngx_http_request_t *r = ev->data;
+    --r->main->blocked;
+    ngx_http_finalize_request(r, NGX_DONE);
```
Comment (allanrbo) [inline nginx/modsecurity/ngx_http_modsecurity.c:743]: I see in the version I wrote november of 2018, I had `ngx_http_core_run_phases(r);` here, to let nginx continue to send the request to the backend (like the comment I wrote above says). At what point does that happen today? I thought `ngx_http_finalize_request` would prevent it from going through the rest of the modules pipeline.

Does `ngx_http_finalize_request` evaluate the reference count and free it if 0?

(also had `r->aio = 0`, though I don't remember what it's for anymore...)

File: nginx/modsecurity/ngx_http_modsecurity.c [hunk 5]
```diff
@@ -819,57 +761,72 @@ ngx_http_modsecurity_prevention_task_offload(ngx_http_request_t *r)
...
-    request_rec *req = ctx->req;
 
     // Load request to request rec
     if (ngx_http_modsecurity_load_request(r) != NGX_OK ||
             ngx_http_modsecurity_load_headers_in(r) != NGX_OK) {
         return NGX_ERROR;
     }
 
-    if (modsecIsRequestBodyAccessEnabled(req) && (r->headers_in.content_length || r->headers_in.chunked)) {
+    // ModSecurity only initializes the internal request context upon the call
```
Comment (allanrbo) [inline nginx/modsecurity/ngx_http_modsecurity.c]: If I understand it right, the comment kind of doesn't really blend into the code well once this PR is merged. It's talking about Modsecurity as the old behaviour, right? This is still Modsecurity, and the code right below contradicts the comment, because we initialize the request context right here and now.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 1 above)
Comment (allanrbo) [inline nginx/modsecurity/ngx_http_modsecurity.c:741]: I'm a little confused. It used to be `--` previously too. Now that we're doing `++` an extra time, should it be `-=2`?

**Review by allanrbo — DISMISSED** [review]

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 1 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:741]: Makes sense, I'll add a comment.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 2 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:829]: The count is decremented inside the call to `ngx_http_finalize_request()`. This is the way advertized by Nginx development guide: https://nginx.org/en/docs/dev/development_guide.html#http_request (see the "count" field).

Both `count` and `blocked` belong to the `ngx_http_request_t` structure. All allocations are done from the pool associated with an HTTP request so it is sufficient to guarantee that the request lives long enough because in this case its pool is not destroyed too early.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 3 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:276]: This is how the code looked originally until the PR #109 has been merged in.
It is also how it still looks in the origin: https://github.com/SpiderLabs/ModSecurity/blob/v2/master/nginx/modsecurity/ngx_http_modsecurity.c#L235

I don't mind using `ngx_pstrdup0()` instead though, should not be a big deal. What do you think?

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 4 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:743]: We still call `ngx_http_core_run_phases(r)` in the prevention mode completion handler. This is needed to pass the request through the Nginx chain of handlers again once its offloaded processing is completed.

For detection mode, however, we don't need to do this because we return `NGX_DECLINED` from the handler immediately so the request keeps being handled and that is how we get it sent to the backend immediately without awaiting for the background task to complete.

`ngx_http_finalize_request()` is the documentation-recommended way to decrement the `count` and make sure the request is properly released when the ref count hits zero: https://nginx.org/en/docs/dev/development_guide.html#http_request

`r->aio` is not well documented, however, from what I could gather it effectiely tells Nginx there is some ongoing async I/O operation on this request so it does not fulfill the request. For prevention mode that was fine to do, though not sure if that was necessary but wouldn't hurt anyway. Here, it would cause the request to not be sent to the backend which is not exactly what we want.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 5 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c]: Good point. Let me rephrase this to make it clear why we need this API and the call to it.

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 1 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:741]: It is incremented once for a detection mode and once for a prevention mode. These modes have different completion handlers. We have been decrementing `blocking` for prevention mode for a long time now but never before used `count` and `blocking` for detection mode. Now we do, so this is why the line above has been added.

**Review by azhao155 — DISMISSED** [review]

File: nginx/modsecurity/ngx_http_modsecurity.c (hunk 3 above)
Comment (vkrivopalov) [reply nginx/modsecurity/ngx_http_modsecurity.c:276]: I found that the comment a few lines above explains why it is safe to use `path.data`.

**Review by allanrbo — APPROVED** [review]
